File: app/src/main/java/com/kode/utair/presentation/ui/adapters/DailyForecastAdapter.java
package com.kode.utair.presentation.ui.adapters;

import android.content.Context;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.kode.utair.R;
import com.kode.utair.domain.models.WeatherForecastEntity;

import java.text.SimpleDateFormat;
import java.util.List;

public class DailyForecastAdapter extends RecyclerView.Adapter<DailyForecastAdapter.DayForecastViewHolder> {

    private Context context;
    private LayoutInflater layoutInflater;
    private List<WeatherForecastEntity.DailyForecast> dailyForecasts;

    private SimpleDateFormat dateFormatter = new SimpleDateFormat("d MMM, E");

    public DailyForecastAdapter(Context context, List<WeatherForecastEntity.DailyForecast> dailyForecasts) {
        this.context = context;
        this.layoutInflater = LayoutInflater.from(context);
        this.dailyForecasts = dailyForecasts;
    }

    @Override
    public DayForecastViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = layoutInflater.inflate(R.layout.item_daily_forecast, parent, false);
        return new DayForecastViewHolder(view);
    }

    @Override
    public void onBindViewHolder(DayForecastViewHolder holder, int position) {
        WeatherForecastEntity.DailyForecast dailyForecast = dailyForecasts.get(position);

        holder.dayLabel.setText(dateFormatter.format(dailyForecast.getDate().toDate()));
        holder.hourlyForecastList.setAdapter(new HourlyForecastAdapter(context, dailyForecast.getHourlyForecasts()));
    }

    @Override
    public int getItemCount() {
        return dailyForecasts.size();
    }

    public class DayForecastViewHolder extends RecyclerView.ViewHolder {

        private TextView dayLabel;
        private RecyclerView hourlyForecastList;

        private DayForecastViewHolder(View itemView) {
            super(itemView);

            this.dayLabel = (TextView) itemView.findViewById(R.id.day_label);
            this.hourlyForecastList = (RecyclerView) itemView.findViewById(R.id.hourly_forecast_list);

            hourlyForecastList.setLayoutManager(new LinearLayoutManager(itemView.getContext(), LinearLayoutManager.HORIZONTAL, false));
        }

    }
}

File: core/src/main/java/com/utair/presentation/ui/weatherforecast/DailyForecastAdapter.kt
package com.utair.presentation.ui.weatherforecast

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.utair.core.R
import com.utair.domain.global.models.WeatherForecast.DailyForecast
import com.utair.presentation.ui.weatherforecast.DailyForecastAdapter.DayForecastViewHolder
import kotlinx.android.synthetic.main.item_daily_forecast.view.*
import java.text.SimpleDateFormat

class DailyForecastAdapter(
        private val context: Context,
        private val dailyForecasts: List<DailyForecast>
) : RecyclerView.Adapter<DayForecastViewHolder>() {

    private val layoutInflater: LayoutInflater = LayoutInflater.from(context)
    private val dateFormatter = SimpleDateFormat("d MMM, E")

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): DayForecastViewHolder {
        val view = layoutInflater.inflate(R.layout.item_daily_forecast, parent, false)
        val layoutManager = LinearLayoutManager(view.context, LinearLayoutManager.HORIZONTAL, false)
        view.hourlyForecastList.layoutManager = layoutManager
        return DayForecastViewHolder(view)
    }

    override fun onBindViewHolder(holder: DayForecastViewHolder, position: Int) {
        val dailyForecast = dailyForecasts[position]
        val itemView = holder.itemView
        itemView.dayLabel.text = dateFormatter.format(dailyForecast.date.toDate())
        itemView.hourlyForecastList.adapter = HourlyForecastAdapter(context, dailyForecast.getHourlyForecasts())
    }

    override fun getItemCount() = dailyForecasts.size

    class DayForecastViewHolder(itemView: View) : RecyclerView.ViewHolder(itemView)

}